fix: IsAuthenticated cancellation races

3v1n0 opened the pull request:
Both on the client side and the server side we have some troubles in handling repeated `IsAuthenticated` requests followed by cancellations.

In particular:
 - The client didn't wait for previous authentication/cancellation to be completed to start a new one
 - The server may end up sending the cancellation events to the broker when the Authentication wasn't started yet
 - The client may send a cancelled event to the authentication model while the broker is still in the process of cancelling a previous request, and the server may still forward it

See commits fixing (or workarounding) each issue here for deeper explanations and the rationale for the fix or workaround in place.

UDENG-3417

didrocks reviewed (approved):
Thanks for moving the check on the PAM side! I didn’t spot anything hairy and I think the new tests makes sense.

In the end, it even looks less hackish than the first PR version :D

We definitively need to take this is account into the V2 design.

3v1n0 commented:
> In the end, it even looks less hackish than the first PR version :D

Yeah, some parts are, the fact that we've to wait for some unknown amount of time doesn't make me fully happy, but we still had something similar before and at least now everything is in a single place.

Meanwhile, rebased and dropped the already reverted commits (so no-change force-push in the end).

3v1n0 commented on examplebroker/broker.go:
```diff
@@ -408,6 +410,24 @@ func getPasswdResetModes(info sessionInfo, supportedUILayouts []map[string]strin
+	baseCode := 1337
+	qrcodeURIs := []string{
+		"https://ubuntu.com",
+		"https://ubuntu.fr/",
```
We are all represented now :)
